Filter Roomserver Events Based on Application Service

anoadragon453 opened the pull request:
When Application Services are registered, they specify a number of namespaces (regexes based on userID, RoomID and Room Alias) for events that they want to know about. We pull these namespaces in at startup and compile their regex into Regexp objects, which we later use to quickly check if several properties of an incoming event match what a given application service wants.

Checking sender names and room IDs were quite simple as they are part of the event, however to get all the aliases of a room, a new aliasAPI was necessary. `GetAliasesForRoomID` was added for this, right alongside `GetRoomIDForAlias` which was renamed from its earlier quite vague name of `GetAliasRoomID`.

I believe this new alias API is the correct way to get aliases for a given RoomID, though a database query for Aliases from RoomID was already available, so let me know if this was already possible somehow and the new call isn't necessary.

At the moment it just prints out whether an app service was interested in an event or not, but the next step is to queue up that event and have a module batch and send them off to the appservice.

anoadragon453 commented:
Ok, should be filtering all known events now.

erikjohnston commented on src/github.com/matrix-org/dendrite/appservice/consumers/roomserver.go:
```diff
@@ -142,3 +156,57 @@ func (s *OutputRoomEventConsumer) lookupStateEvents(
+	// Check all known room aliases of the room the event came from
+	queryReq := api.GetAliasesForRoomIDRequest{RoomID: event.RoomID()}
+	var queryRes api.GetAliasesForRoomIDResponse
+	if err := s.alias.GetAliasesForRoomID(context.TODO(), &queryReq, &queryRes); err == nil {
```
You should be threading the context through. `context.TODO()` should only be used when you otherwise don't know quite where to get one yet, but we already have made a context for `UpdateMemberships`, so lets use that one.

anoadragon453 replied:
Gotcha. Still figuring out how context works :)

erikjohnston commented on src/github.com/matrix-org/dendrite/appservice/consumers/roomserver.go:
```diff
@@ -142,3 +156,57 @@ func (s *OutputRoomEventConsumer) lookupStateEvents(
+			}
+		}
+	} else {
+		log.WithError(err).Errorf("Unable to get aliases for Room with ID: %s", event.RoomID())
```
Probably better to use the structured logging, so something like:

```golang
log.WithFields(log.Fields{
    "room_id": event.RoomID(),
}).WithError(err).Error("Unable to get aliases for room")
```

erikjohnston commented on src/github.com/matrix-org/dendrite/common/config/appservice.go:
```diff
@@ -182,9 +199,9 @@ func checkErrors(config *Dendrite) error {
 
-	return nil
+	err = setupRegexps(config)
+	return err
```
Why are we not just `return setupRegexps(config)`, and why do we need to name the return value?

anoadragon453 replied:
Derp, yes that really should be `return setupRegexps(config)`.

erikjohnston commented on src/github.com/matrix-org/dendrite/common/config/config.go:
```diff
@@ -235,12 +235,19 @@ type Dendrite struct {
+		ExclusiveApplicationServicesAliasRegexp *regexp.Regexp
+		// TODO: Huh? When a room ID is created, we block the client from
+		// creating it if it falls under an application service's exclusive
+		// roomID regex? But why tho
```
We should probably find out the answer to this question before writing the code and landing it :)

erikjohnston commented on src/github.com/matrix-org/dendrite/roomserver/alias/alias.go:
```diff
@@ -98,14 +98,30 @@ func (r *RoomserverAliasAPI) GetAliasRoomID(
+
+	response.Aliases = aliases
+	return nil
+}
```
Where was the implementation for `GetAliasesFromRoomID`? I'm surprised I'm not seeing that removed in the diff.

anoadragon453 replied:
There wasn't one. I've had to add it :)

erikjohnston replied:
Then how was `RoomserverAliasAPIDatabase` implemented?

anoadragon453 replied:
The database queries and supporting functions were already there, a method to expose it to the RoomserverAliasAPI was all that was needed.

anoadragon453 commented:
* Prevented user from creating more than one regex per namespace type

  - Got caught out by realizing I had an extra '-' in the config file. This
prevents anyone from making the same mistake :)

* There was a pointer reference to a string slice. Slices are already
pointers, so it wasn't necessary. Testing shows this hasn't broken
anything.

* Removed exclusive RoomID namespace regex, as we won't need to check
upon room creation if the ID is reserved exclusively by an AS (as this
is silly and horribly inefficient).

* Fixed all else mentioned

@erikjohnston Back to you
